packages/realm/src/Realm.ts

```diff
@@ -511,16 +511,24 @@ export class Realm {
       );
       const table = binding.Helpers.getTable(this.internal, objectSchema.tableKey);
       table.removeObject(obj.key);
+    } else if (subject instanceof List) {
+      const internalList = subject.internal;
+      internalList.removeAll();
+    } else if (subject instanceof Results) {
+      const internalResult = subject.internal;
+      internalResult.clear();
     } else if (Array.isArray(subject) || Symbol.iterator in subject) {
-      // TODO: Optimize this to not get the helper on every iteration
-      for (const object of subject) {
-        assert.instanceOf(object, RealmObject);
-        const { objectSchema } = this.classes.getHelpers(object);
-        const table = binding.Helpers.getTable(this.internal, objectSchema.tableKey);
-        table.removeObject(object[INTERNAL].key);
+      if (subject.length > 0) {
+        // Get the objectSchema from the first element, and assume that all elements are of the same object type
+        const { objectSchema } = this.classes.getHelpers(subject[0]);
```

Comment:
This needs to happen after the assert that `object` is instance of `RealmObject`.

I actually think it would be better to leave it as it was and simply remove the TODO. It will be a lot less hot code now that you've added native support for `List` and `Result` and getting the helper on every object makes the API more useful as you can call it with an array of different objects.

Reply:
I could also assert `object` at that level as well.

Reply:
```typescript
        assert.instanceOf(subject[0], RealmObject);
        const { objectSchema } = this.classes.getHelpers(subject[0]);
        for (const object of subject) {
          const table = binding.Helpers.getTable(this.internal, objectSchema.tableKey);
          table.removeObject(object[INTERNAL].key);
        }
```

Reply:
But if for some reason, one was to just wildly put a bunch of realm objects of a different type into an array and try and delete it, this might cause some issues.  Maybe just putting it back the way it was is fine